spec/views/jwt_api_entreprise_mailer/satisfaction_survey.html.erb_spec.rb

```diff
@@ -0,0 +1,66 @@
+require 'rails_helper'
+
+RSpec.describe 'rendering the mail template' do
+  before do
+    assign(:full_name, full_name)
+    render template: "jwt_api_entreprise_mailer/satisfaction_survey.#{format}.erb"
+  end
+
+  let(:embedded_form_truncated_content) do
+    <<~EMBEDDED.chomp
+      quelle est votre satisfaction</strong> concernant votre
+    EMBEDDED
+  end
+
+  let(:link_to_the_form) do
+    <<~LINK.chomp
+      <a href="https://startupdetat.typeform.com/to/bFo1H9NJ">Répondre à l&#39;enquête utilisateur</a>
+    LINK
+  end
+
+  subject do
+    rendered
+  end
+
+  context 'when in HTML' do
+    let(:format) { 'html' }
+
+    context 'with a name' do
+      let(:full_name) { 'Bob MARTIN' }
+
+      it { expect(subject).to match /Bonjour Bob MARTIN,/ }
+    end
+
+    context 'without a name' do
+      let(:full_name) { nil }
+
+      it { expect(subject).to match /Bonjour,/ }
+    end
+
+    after do
+      expect(subject).to match embedded_form_truncated_content
+      expect(subject).not_to match link_to_the_form
+    end
+  end
+
+  context 'when in plaintext' do
+    let(:format) { 'text' }
+
+    context 'with a name' do
+      let(:full_name) { 'Bob MARTIN' }
+
+      it { expect(subject).to match /Bonjour Bob MARTIN,/ }
+    end
+
+    context 'without a name' do
+      let(:full_name) { nil }
+
+      it { expect(subject).to match /Bonjour,/ }
+    end
+
+    after do
+      expect(subject).not_to match embedded_form_truncated_content
+      expect(subject).to match link_to_the_form
+    end
+  end
+end
```

Comment:
J'aurais tendance à dire que ces specs font doublons avec celles des mailers où la présence du contenu important est déjà testée.

Reply:
D'accord avec toi, ça fait doublon :/

A propos, les tests de mails sont un peu comme des tests de controllers : on peut très bien y tester en plus les vues qui y sont générées. Mais ça a l'inconvéniant d'augmenter le périmètre des testes, et d'augmenter le volume du code pour _tout_ tester.

Ma proposition pour limiter les doublons, c'est plutôt qu'on se mette à exploiter les tests de type [view](https://relishapp.com/rspec/rspec-rails/v/5-0/docs/view-specs) lorsqu'on teste des vues. C'est dans RSpec pour ça, et ça aide au maintien de specs courte et précise.

Pour info, en relisant les testes de la vue de l'email à l'intérieur des tests du mailer, je viens de remarquer un faux positif. Comme quoi, c'est pas assez clair et ça ajoute de la complexité. Voici le code :

```ruby
it 'contains the link to the form' do
  token_url = "https://startupdetat.typeform.com/to/bFo1H9NJ"

  expect(subject.html_part.decoded).to include(token_url)
  expect(subject.text_part.decoded).to include(token_url)
end
```

Le teste passe, bien que dans la version actuelle du code le contenu des emails au format HTML ne contient pas de lien. Il contient plutôt des assets qui pointent vers "`https://startupdetat.typeform.com/to/bFo1H9NJ`". Bref, je vais fixer ça, en vous proposant si vous êtes d'accord de privilégier les tests de mailer pour déterminer le bon comportement propre aux mailers (le sujet, la source, le destinataire, etc), et pour l'avenir de privilégier les tests de vue pour tester à proprement parler ce qui concerne les vues.

Reply:
Comme les seules vues à proprement parler que l'on a dans le projets sont des mails et que les "tests" de vues sont uniquement là pour s'assurer de la présence de liens ou données clés des mails transactionnels (ce que je ne considère pas vraiment comme des tests de vues du coup) qui sont envoyés je ne suis pas certain de la plus value de séparer ces tests.

Reply:
Le fait de ne tester que la présence des données clés permet surtout de ne pas casser la suite de test dès la reformulation ou mise en forme des emails qui sont envoyés.

Reply:
Oui, c'est pratique et ça fonctionne.

Après, même si les vues actuelles sont simples, l'avantage de privilégier l'emploi de specs de vues c'est de séparer les types de tests en fonction des types de logique, et d'avoir en fonction de ça des matchers plus adéquats. Ça limite aussi la complexité en permettant de chirurgicalement pouvoir n'avoir à injecter que le strict nécessaire pour la bonne instanciation du test voulu. Par exemple si je veux tester une vue, je n'ai pas besoin de savoir si elle vient d'un mailer ou d'un SMS ou d'un controller classique, je n'ai pas besoin de renseigner dans mon test un sujet ni l'email d'un destinataire non plus. Juste je déclare les variables nécessaires à la génération de la template, et j'y teste ce qui est pertinant.

Un dernier point,

> Le fait de ne tester que la présence des données clés permet surtout de ne pas casser la suite de test dès la reformulation ou mise en forme des emails qui sont envoyés.

Je suis d'accord là aussi, mais je voudrais nuancer : le fait qu'un test casse à cause d'un changement dans une template n'est pas forcément mauvais, c'est même mieux parfois, notamment pour avoir une vision sur l'intégralité de nos emails. C'est surtout utile pour les templates complexes, je le reconnais. Mais nous, malgré la simplicité des notres, nous générons par exemple actuellement des documents HTML complets qui sont embarqués dans une layout HTML complète. En testant le rendu d'au moins un email, cette (petite) erreur ne serait pas arrivée. Note : je pourrai proposer une PR distincte pour fixer ça.

Reply:
> Note : je pourrai proposer une PR distincte pour fixer ça.

Commençons par une issue pour en discuter :D 

Mon avis (très perso) est que c'est overkill, et qu'en pratique ça va casser tout le temps pour rien (imaginons Dorine change les wordings).

Reply:
Je suis totalement d'accord.

Ça doit être mon message n'était pas assez clair, mais la petite note finale que j'avais ajoutée ne concernait que l'erreur de générer des pages HTML imbriquées dans des pages HTML, comme ça par exemple (c'est le rendu HTML de `JwtApiEntrepriseMailer#creation_notice`) :

```html
<!DOCTYPE html>
<html>
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <style>
      /* Email styles need to be inline */
    </style>
  </head>

  <body>
    <!DOCTYPE html>
<html>
  <head>
    <meta content='text/html; charset=UTF-8' http-equiv='Content-Type' />
  </head>
  <body>
    <p>Bonjour,</p>

    <p>Un nouveau token est disponible dans votre espace client.<p>

    <p>Vous pouvez le récupérer en suivant ce lien : https://sandbox.dashboard.entreprise.api.gouv.fr/me/tokens/</p>

    <p>Ce token possède les droits d'accès suivants :</p>
    <ul>
        <li>Role 1</li>
        <li>Role 2</li>
        <li>Role 3</li>
        <li>Role 4</li>
    </ul>


    <p>N'oubliez pas, vous pouvez consulter les statistiques d'utilisations propre à ce jeton en temps réel. Voici les statistiques actuellement disponibles:</p>

    <ul>
      <li>la liste des 50 derniers appels</li>
      <li>les pourcentages des différents code HTTP sur les dernières 30h/8j</li>
      <li>le nombre d'appels par API sur les dernières 30h/8j</li>
    </ul>

    <p>Si vous avez besoin d'un nouveau token contactez-nous : <a href="mailto:support@entreprise.api.gouv.fr">support@entreprise.api.gouv.fr</a></p>

    <p>Vous recevez cet email car vous êtes renseigné parmi les contacts pour le compte user_1@example.org. Les accès se font uniquement à partir de cette adresse email.</p>

    <p>Cordialement,</p>
    <p>L'équipe API Entreprise</p>
  </body>
</html>

  </body>
</html>
```

Je pense qu'objectivement nous sommes tous d'accord pour dire que ce n'est pas bien, sans avoir à en débattre.

Reply:
Ta remarque fait echo avec mon commentaire sur cette vue on je dis justement qu'il faut refactoriser dans le mailer.html.erb

Reply:
Je propose de faire ça via une autre PR.